anvil/src/drawing.rs

```diff
@@ -1,158 +1,126 @@
+    R: Renderer<TextureId = T> + ImportAll,
 {
-    fn id(&self) -> usize {
-        0
-    }
-
-    fn location(&self, scale: impl Into<Scale<f64>>) -> Point<f64, Physical> {
-        self.position.to_f64().to_physical(scale)
-    }
-
-    fn geometry(&self, scale: impl Into<Scale<f64>>) -> Rectangle<i32, Physical> {
-        Rectangle::from_loc_and_size(self.position, self.size).to_physical_precise_round(scale)
-    }
-
-    fn accumulated_damage(
-        &self,
-        scale: impl Into<Scale<f64>>,
-        _: Option<SpaceOutputTuple<'_, '_>>,
-    ) -> Vec<Rectangle<i32, Physical>> {
-        let scale = scale.into();
-        vec![Rectangle::from_loc_and_size(self.position, self.size).to_physical_precise_up(scale)]
-    }
-
-    fn opaque_regions(&self, _scale: impl Into<Scale<f64>>) -> Option<Vec<Rectangle<i32, Physical>>> {
-        None
-    }
-
-    fn draw(
-        &self,
-        _renderer: &mut R,
-        frame: &mut <R as Renderer>::Frame,
-        scale: impl Into<Scale<f64>>,
-        location: Point<f64, Physical>,
-        _damage: &[Rectangle<i32, Physical>],
-        _log: &Logger,
-    ) -> Result<(), <R as Renderer>::Error> {
-        let scale = scale.into();
-        frame.render_texture_at(
-            &self.texture,
-            location.to_i32_round(),
-            1,
-            scale,
-            Transform::Normal,
-            &[Rectangle::from_loc_and_size(
-                (0, 0),
-                self.size.to_physical_precise_round(scale),
-            )],
-            1.0,
-        )?;
-        Ok(())
+    type RenderElement = PointerRenderElement<R>;
+    fn render_elements<E>(&self, location: Point<i32, Physical>, scale: Scale<f64>) -> Vec<E>
+    where
+        E: From<PointerRenderElement<R>>,
+    {
+        match &self.status {
+            CursorImageStatus::Hidden => vec![],
+            CursorImageStatus::Default => {
+                if let Some(texture) = self.texture.as_ref() {
+                    vec![
+                        PointerRenderElement::<R>::from(TextureRenderElement::from_texture_buffer(
+                            location.to_f64(),
+                            texture,
+                            None,
+                            None,
+                        ))
+                        .into(),
+                    ]
+                } else {
+                    vec![]
+                }
+            }
+            CursorImageStatus::Surface(surface) => {
+                let elements: Vec<PointerRenderElement<R>> =
+                    smithay::backend::renderer::element::surface::render_elements_from_surface_tree(
+                        surface, location, scale,
+                    );
+                elements.into_iter().map(E::from).collect()
+            }
+        }
     }
 }
 
 #[cfg(feature = "debug")]
 pub static FPS_NUMBERS_PNG: &[u8] = include_bytes!("../resources/numbers.png");
 
 #[cfg(feature = "debug")]
+#[derive(Clone)]
```

Comment:
I

Reply:
Yeah, cloning the commit counter and the id is not really efficient, but they probably get redrawn a bunch anyway and it's only in debug mode... but I suppose we could fix this?

Reply:
sure